### PR TITLE
irmin: add support for multi-version objects

### PR DESCRIPTION
**Pull request by samoht**
First tentative to implement #1439 

Currently:
- only multi-version nodes are supported (we should do the same for contents).
- the code compiles but the tests are broken
- there's various cleanup patches that can be separated in tree.ml and upstreamed directly
- we need a more clever migration function in Tree.ml as the current one is partial (and probably broken)
- more tests tests tests

### CODE REVIEW

**Comment by Ngoguey42** [comment]
I've investigated this UT fail:
```
┌──────────────────────────────────────────────────────────────────────────────┐
│ [FAIL]        MEM                  8   Tree caches and hashconsing.          │
└──────────────────────────────────────────────────────────────────────────────┘
+258739us irmin           [DEBUG] set /a/b
+258785us irmin.mem       [DEBUG] find master
+258811us irmin           [DEBUG] Head.find -> 
+258830us irmin.tree      [DEBUG] Tree.add_tree /a/b
+258865us irmin.mem       [DEBUG] mem f29a17af8c40ffaa06d650a5e9b8b3508a635774
+258885us irmin.mem       [DEBUG] mem 73e5f562523249e0df21c98d9ab9c05cc3ce001d
+258902us irmin.mem       [DEBUG] add -> 0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33
+258918us irmin.mem       [DEBUG] add -> 73e5f562523249e0df21c98d9ab9c05cc3ce001d
+258934us irmin.mem       [DEBUG] add -> f29a17af8c40ffaa06d650a5e9b8b3508a635774
+258951us irmin.tree      [DEBUG] Tree.export -> f29a17af8c40ffaa06d650a5e9b8b3508a635774
+258968us irmin.mem       [DEBUG] add -> e0dd105c31fc2408d05fa840f4e8230b9269f697
+258983us irmin.mem       [DEBUG] test_and_set
+258998us irmin.mem       [DEBUG] find master
+259016us irmin.mem       [DEBUG] find master
+259029us irmin.mem       [DEBUG] find e0dd105c31fc2408d05fa840f4e8230b9269f697
+259045us irmin           [DEBUG] Head.find -> e0dd105c31fc2408d05fa840f4e8230b9269f697
+259059us irmin.tree      [DEBUG] Tree.find_tree 
ASSERT inspect
+259083us irmin.tree      [DEBUG] Tree.add /foo
+259097us irmin.mem       [DEBUG] find f29a17af8c40ffaa06d650a5e9b8b3508a635774
ASSERT inspect:0
ASSERT val-v:0
+259136us irmin.tree      [DEBUG] Tree.add /bar/foo
ASSERT inspect:1
ASSERT val-v:1
ASSERT val-list:1
+259175us irmin.tree      [DEBUG] Tree.hash
ASSERT inspect:2
ASSERT val-v:2
+259221us irmin.mem       [DEBUG] clear
+259239us irmin.mem       [DEBUG] clear
+259254us irmin.mem       [DEBUG] clear
+259270us irmin.mem       [DEBUG] clear
+259287us irmin.mem       [DEBUG] close
+259303us irmin.mem       [DEBUG] close
+259318us irmin.mem       [DEBUG] close
+259334us irmin.mem       [DEBUG] close
FAIL val-v:2

   Expected: `0'
   Received: `1'
Raised at Alcotest_engine__Test.check in file "src/alcotest-engine/test.ml", line 191, characters 4-243
Called from Irmin_test__Store.Make.test_tree_caches.test.(fun) in file "src/irmin-test/store.ml", line 1027, characters 6-70
Called from Irmin_test__Common.Make_helpers.run.(fun) in file "src/irmin-test/common.ml", line 157, characters 11-20
Called from Lwt.Sequential_composition.catch in file "src/core/lwt.ml", line 2023, characters 16-20
Re-raised at Irmin_test__Common.Make_helpers.run.(fun) in file "src/irmin-test/common.ml", line 160, characters 36-72
Called from Irmin_test__Common.Make_helpers.run in file "src/irmin-test/common.ml", line 153, characters 6-297
Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 229, characters 17-23
Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
```

This is caused by this counter check https://github.com/mirage/irmin/blob/062ae9a4dce7f1925b90109eaa3f17a69740e9b7/src/irmin-test/store.ml#L1027

In order to build a node containing one entry, we start with the empty node and add it an entry to it. On master this process produces a `Value (repo, empty_node, Some {Add ..})`, on this PR the same process produces a `Map {..}`.

Both are fine IMO.

**Comment by samoht** [comment]
@Ngoguey42 this is fixed by #1443

File: src/irmin/info.ml [hunk 1]
```diff
@@ -16,26 +16,35 @@
...
   type message = string [@@deriving irmin]
-
-  type t = { date : int64; author : author; message : message }
+  type version = V.t [@@deriving irmin]
+
+  type t = {
+    date : int64;
+    author : author;
+    message : message;
+    version : version;
```
Comment (Ngoguey42) [inline src/irmin/info.ml:28]: How is this new `version` field going to play with existing stores?

`V.default` is `unit` so it should be fine in that case, but as soon as we use will use something else than `unit`, the deserialisation will 💥

**Comment by Ngoguey42** [comment]
`dune exec -- ./test/irmin-mem/test.exe test '^MEM' 0,1,2,3,4,5,6,7,8,9,10,11,13` is fine while `dune exec -- ./test/irmin-mem/test.exe test '^MEM' 12,13` fails:

```
  [OK]          MEM                 12   High-level store operations.
> [FAIL]        MEM                 13   High-level operations on trees.
```

I'm not sure why

**Comment by samoht** [comment]
Force-pushed new changes - porting tezos-context-hash to the new API is painful. I've heard the feedback of everyone on that point and we should probably simplify that API again (#1448 is a good way forward but we can also probably simplify the `Make_ext` functors too).

**Comment by Ngoguey42** [comment]
The way I will implement the inodes without stable_hash will depend a lot on the way we support multi hash.

I'll do it as soon as we settle on a solution.

**Comment by mattiasdrp** [comment]
> 
> 
> `dune exec -- ./test/irmin-mem/test.exe test '^MEM' 0,1,2,3,4,5,6,7,8,9,10,11,13` is fine while `dune exec -- ./test/irmin-mem/test.exe test '^MEM' 12,13` fails:
> 
> ```
>   [OK]          MEM                 12   High-level store operations.
> > [FAIL]        MEM                 13   High-level operations on trees.
> ```
> 
> I'm not sure why

When commenting the following lines https://github.com/samoht/irmin/blob/versions/src/irmin-test/store.ml#L954-L984 the test runs ok even when testing `High-level store operations`. I'm not exactly sure why this is happening, I'm investigating it

**Comment by CraigFe** [comment]
Closing as we have no immediate need for multi-version nodes / contents support in `irmin-pack` (but this requirement may pop up again later).

**Pull request closed without merging**